F, d fitting

JoepVanlier opened the pull request:
This provides the fitting backend for the WLC fitting. This backend allows global fitting of multiple datasets simultaneously, keeping some parameters the same between datasets, while varying others. It also includes several WLC models that are typically used in literature. Where possible, models were provided in analytical form.

To improve performance, model Jacobians are provided and propagated for composite and inverted models. For the inverse models where the inverse is not available as an analytical function, the inversion was performed using least_squares (again using the analytical jacobian if supplied). In addition, sparse matrices are used to speed everything up. Tests were added for each of the models as well as the complete fitting infrastructure.

Documentation (a tutorial and three examples) is available on readthedocs (private for now):
https://lumicks-pylake.readthedocs.io/en/fdfit_history/

I would highly recommend going through the tutorial before reading the code to get some idea of the different use cases that we have to support.

An attempt was made to keep the API relatively simple. For example:
```
M = (force_model('DNA', 'WLC') + force_model('DNA', 'offset')).invert()
M.load_data(d, f)
M.load_data(d, f, DNA_Lp="DNA_Lp2")  # Creates a new independent parameter DNA_Lp2 for the second dataset, all others are shared.
F = FitObject(M)

F.fit()
plt.figure()
F.plot_data()
F.plot_model()
```

Out of scope (for now), but which would be nice to have in the future:
- API/Interface straight from fdcurve into load_data.
- Additional / better data slicing options.

JoepVanlier commented:
> This an impressive tool and overall is amazing already. I was able to fit multiple curves easily and in very little time. This is my feedback, is not really a code review but I hope it could be useful.

Thanks a lot for using it from a user point of view. I really appreciate it. Lots of good points in here.

> 1. subtract_offset throws an error (I believe the documentation is not correct)
>    ```
>     odijk = pylake.inverted_odijk("DNA").subtract_offset("d_offset")
>     ---------------------------------------------------------------------------
>     AttributeError                            Traceback (most recent call last)
>     <ipython-input-39-8d90f6071f4c> in <module>
>     ----> 1 odijk_with_offset = pylake.inverted_odijk("DNA").subtract_offset("d_offset")
>     
>     AttributeError: 'Model' object has no attribute 'subtract_offset'
>    ```

Correct. I have added a commit that fixes this.

> 2. `parameter_names` should be `parameters_name` in my opinion

Why? Parameter_names is English, whereas parameters_name is not. I'd sooner expect the former personally. I don't have a strong preference, so I don't mind changing it, but I'm just curious what gives a preference for the former.

> 3. if I invert the model with an `force_offset` it becomes effectively a `distance_offset`, but keeps the name, could this create some confusion?

I don't think it changes to be honest. In `(M1+force_offset).invert()`, offset is still a force offset. `M1.invert() + distance_offset` is a distance offset however. I do agree that it is somewhat confusing having a single 'offset' though. I've made explicit `force_offset` and `distance_offset` models now, so that users have to be explicit and end up with more readable compositions. There is also validation of compatibility of dependent and independent variable now (added in the last PRs), which means that you can't add incompatible models anymore.

> 4. Probably out of scope of this PR but before the official release I would be nice to implementing loading fd curve items:
>    this would be nice:
>    ```
>     fd3 = file.fdcurves['3']
>     data3 = odijk_fwd.load_data(fd3)
>     data3 = odijk_fwd.load_data(fd3.d,fd3.f,name='3')
>    ```
>    
>    
>    instead the current form is a bit verbose IMO:
>    ```
>     data3 = odijk_fwd.load_data(fd3.distance2.data,fd3.f.data,name='3')
>    ```

Definitely. I would say that this is probably a good one to combine with the better slicing functionalities, since then we have to really think about how to slice fds in the first place. This will become a big issue when doing things with many breaks/refolds, like the AdK data.

> 5. `print(model)` or `help(model)` could print the equation and the parameters with units and defaults.

Great idea. Done. You now get a little model overview, optionally formatted in a table and latex equation if you're in a jupyter notebook.

> 6. Which data is dependent vs indipendent in the fit is error-prone, having `load_force` and `load_distance` and then do the right thing based on whether the model is direct or inverse would be nice.

I agree that this is a point of difficulty. I'd suggest a different solution since I think splitting loading over two different calls is more error prone (you can end up loading incompatible sets) than having it in a single one.

I suggest changing the data loading to a named function instead. We can force named arguments directly and match them with the dependent and independent variable.

In the last PRs I have derived Fd versions of the models from the base models, but which have a customized load_data that takes an f and d, and selects the appropriate one for the independent and dependent variables.

So all calls will be:
`model.load_data(f=force, d=distance, name=name, parameter changes)`

And they will always look the same, independent of whether it is a forward or backward model. The model will figure out which is dependent and independent. I also force use of keywords, meaning that you always have to invoke the f= and d=, to prevent mistakes.

> 7. Some models are much slower, this should be somehow more intuitive, for example
>    ```
>     (pylake.odijk("DNA") + pylake.offset("f")).invert()
>    ```
>    
>    
>    is much slower (at least 30x in my case) than
>    ```
>      pylake.inverted_odijk("DNA").subtract_offset("f_offset") 
>    ```
>    
>    
>    but there is no hint for the user.

Correct. But it is quite clearly described in the docs and there are two examples in the docs which state the difference.

The difficulty is that without adding a lot of infrastructure (where you actually start traversing the dependency graph to see if it's a single forward model), it is not easy to detect legitimate and illegitimate uses of the forward WLC model. Sometimes you have to use the slow version, for example when you have two WLC models added; because no analytic inverse exists for that.

> 8. Tooling for selecting distance and force value ranges for the fit would be nice.

Yes, I fully agree, but I would argue that this ties in the with the slicing thing.

> 9. It is not clear how to fix fit parameters values

`F.parameters["DNA_Lc"].vary = False`

I will add an explicit remark in the docs and the docstring for parameters.

agom commented:
> Why? Parameter_names is English, whereas parameters_name is not. I'd sooner expect the former personally. I don't have a strong preference, so I don't mind changing it, but I'm just curious what gives a preference for the former.

It's just a minor comment, you don't need to change it, it's just that `parameter_names` sounds like a parameter has multiple names, but maybe it's just me. Don't worry about it.

dean0x7d commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,314 @@
+which dataset is force and which is distance via keyed arguments (this is intended to prevent
+mistakes in argument order)::
+
+    data1 = odijk_with_offsets.load_data(f=force1, d=distance1, name="Control")
```
What can the user do with the returned `data1` object? It's not used in the example. I would either remove it or show an example. If it's advanced usage, you can remove the variable here and put it in a separate section on this page.

JoepVanlier replied:
It is used in the example down the road. It's used to pick parameters from the `Fit` as `odijk_fit[data1]`. Do you think we should try and put those closer together? Then I would probably duplicate the earlier code down the road.

dean0x7d commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,314 @@
+passing it one or more models. In return, we get an object we can interact with, which in this
+case we store in `odijk_fit`::
+
+    odijk_fit = pylake.FitObject(odijk_with_offsets)
```
Obligatory naming comment: the "object" part in `FitObject` is kinda redundant. Most things are objects in Python. `Fit` or `Fitter` would be fine.

JoepVanlier replied:
👍  Yes. I've renamed it to `Fit` :)

dean0x7d commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,314 @@
+tend to have reasonable initial guesses and bounds in pylake, but we can set our initial guess and
+a lower and upper bound as follows::
+
+    odijk_fit.parameters["DNA_Lp"].value = 50
+    odijk_fit.parameters["DNA_Lp"].lb = 39
+    odijk_fit.parameters["DNA_Lp"].ub = 80
```
Parameters seem to be the main use of this class. Can we just give it a `__getitem__` to simplify things?
```suggestion
    odijk_fit["DNA_Lp"].value = 50
    odijk_fit["DNA_Lp"].lb = 39
    odijk_fit["DNA_Lp"].ub = 80
```

JoepVanlier replied:
Yep, definitely more beautiful. 👍

dean0x7d commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,314 @@
+We can see that there are no parameters to be fitted. The reason for this is that
+we did not add any data to the model yet. Let's add some and fit this data::
+
+    >>> data1 = odijk_inv.load_data(f=f1, d=d1, name="Control")
```
Nearly all of the examples give data a name. Why not make `name` a required parameter?

JoepVanlier replied:
Yeah. Makes sense.

dean0x7d commented on lumicks/pylake/fitting/parameters.py:
```diff
@@ -0,0 +1,170 @@
+from tabulate import tabulate
+
+
+class Parameter:
+    def __init__(self, value=0.0, lb=-np.inf, ub=np.inf, vary=True, init=None, shared=False, unit=None):
+        """Model parameter
+
+        Parameters
+        ----------
+        value: float
+        lb, ub: float
+        vary: boolean
+        init: float
+        shared: boolean
+        """
+        self.value = value
+        self.lb = lb
+        self.ub = ub
+        self.vary = vary
+        self.shared = shared
+        self.unit = unit
```
I feel like this part is quite under-documented. And I feel like, if this part of the API is not going to be used very often, then it would be better to have more expressive names, e.g. `lb` -> `lower_bound`. The latter would be much clearer for someone who just happened to see this in someone else's code. In general, I think that name length should be inversely proportional to how often the thing is used.

JoepVanlier replied:
True. I've added a little extra docstring for parameters, and changed lb/ub to lower_bound and upper_bound.

JoepVanlier commented on docs/examples/fd_fitting_basic/index.rst:
```diff
@@ -0,0 +1,113 @@
+
+    :nbexport:`Download this page as a Jupyter notebook <self>`
+
+First we import the necessary libraries::
+
+    import matplotlib.pyplot as plt
+    from lumicks import pylake
+    import numpy as np
+    import h5py
```
👍

JoepVanlier commented on docs/examples/fd_fitting_basic/index.rst:
```diff
@@ -0,0 +1,113 @@
+Load the data from disk
+-----------------------
+
+This is pre Bluelake data, therefore we have to extract the data from the h5
```
I'm not entirely sure to be honest. This was data from Onno's F, d package, so I wouldn't be surprised if it was from pre-lumicks. I used it as test for the tWLC. I mean, we could do two things. Omit the example. Repack the data in a faux bluelake format. Then again, we don't ship the data with the examples, so I could just adjust the example as though it was in a bluelake format. Maybe that's the best option.

dean0x7d replied:
It would be ideal to move non-Bluelake data to a separate section and include pure `pylake` examples here.

JoepVanlier replied:
Yes. I mean, I didn't really intend for these to be docs in the first place. Originally they were Jupyter notebooks on the PR. As an alternative, I suggest just leaving the examples out completely and just have the tutorial, since we have no actual Bluelake data for them. After May fourth, maybe we can add the RecA data.

The tWLC is old Onno data, the RecA is old Jordi data and the AdK relies on preprocessing with a piezo script that's not part of pylake and not ready for release.

JoepVanlier commented on docs/examples/fd_fitting_basic/index.rst:
```diff
@@ -0,0 +1,113 @@
+----------------
+
+Set up a twistable worm like chain model with a distance and force offset. By
+default, the twistable_wlc model provided with pylake is defined as distance
```
Good point. 👍

JoepVanlier commented on docs/examples/fd_fitting_intermediate/index.rst:
```diff
@@ -0,0 +1,297 @@
+First we import our required libraries::
+
+    import matplotlib.pyplot as plt
+    import lumicks.pylake as pylake
```
👍

JoepVanlier commented on docs/examples/fd_fitting_intermediate/index.rst:
```diff
@@ -0,0 +1,297 @@
+    import lumicks.pylake as pylake
+    import numpy as np
+
+Load the data from text files
+-----------------------------
+
+The next step is loading the data. In this case, the data is provided as
+text files::
```
Yes, I agree. This is the data that Loïc will regenerate for me with bluelake on 4th of May. So hopefully, if the experiment goes right, we can put bluelake data in its stead.

JoepVanlier commented on docs/examples/fd_fitting_intermediate/index.rst:
```diff
@@ -0,0 +1,297 @@
+For this we want to use an inverted WLC model with an offset in force
+and distance.
+
+While it possible (and equivalent) to generate this model via
```
Ooh good catch. I always read over missing words like that. I should probably install a grammar tool for when doing docs.

JoepVanlier commented on docs/examples/fd_fitting_intermediate/index.rst:
```diff
@@ -0,0 +1,297 @@
+and distance.
+
+While it possible (and equivalent) to generate this model via
+``pylake.force_model('DNA', 'WLC').invert()``, an optimized inverted WLC
```
Yes. You are correct. I don't know how I missed that. Good catch.

JoepVanlier commented on lumicks/pylake/fitting/models.py:
```diff
@@ -0,0 +1,414 @@
+from .model import FdModel
+
+"""
+Available models:
```
Yes, good point. I removed it. The only thing that can happen to that block is that it is forgotten about and goes out of date 💀

JoepVanlier commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,314 @@
+
+    :nbexport:`Download this page as a Jupyter notebook <self>`
+
+Pylake introduces functionality to be able to fit force extension curves with various 
```
True. Changed!

JoepVanlier commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,314 @@
+    :nbexport:`Download this page as a Jupyter notebook <self>`
+
+Pylake introduces functionality to be able to fit force extension curves with various 
+polymer models. This tutorial aims at being a gentle introduction to these fitting 
+capabilities. Each section will first show the more basic mode of use, followed by more
+advanced ways of interacting with the models, data and plots.
```
👍

JoepVanlier commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,314 @@
+
+When fitting data, everything revolves around models. One of these models is the so-called
+extensible worm-like-chain model by Odijk et al. Let's have a look at it. We can construct
+this model using the supplied pylake function odijk.
```
👍

JoepVanlier commented on docs/api.rst:
```diff
@@ -16,3 +16,6 @@ If you are just looking to get started, read the :doc:`/tutorial/index` first.
     correlated_stack.CorrelatedStack
+    fitting.model.Model
+    fitting.fitobject.FitObject
+    fitting.models
```
Ahhh. Alright! Will do! 👍

JoepVanlier commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,314 @@
+parameters in this model. So for instance, the persistence length in this model, would be
+named `DNA_Lp`. Let's have a look at the model and its parameters::
+
+    >>> odijk_forward = pylake.odijk("DNA")
+
+    Model equation:
+
+    d(f) = DNA_Lc * (1 - (1/2)*sqrt(kT/(f*DNA_Lp)) + f/DNA_St)
+
+    Parameter defaults:
+
+    Name      Value  Unit      Fitted      Lower bound    Upper bound
+    ------  -------  --------  --------  -------------  -------------
+    DNA_Lp    40     [nm]      True                  0            inf
+    DNA_Lc    16     [micron]  True                  0            inf
+    DNA_St  1500     [pN]      True                  0            inf
+    kT         4.11  [pN*nm]   False                 0              8
```
Thanks. I've added some more :)

I really like how you have a lot of control over formatting in jupyter notebooks.

JoepVanlier commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,314 @@
+If we also expect an offset in the dependent parameter, we can simply add an offset model to our
+model::
+
+    odijk_with_offsets = pylake.inverted_odijk("DNA").subtract_independent_offset("d_offset") + pylake.force_offset("f")
```
Yes, I agree. Do you have a preferred suggestion?

Originally, `inverted_odijk` was `inv_odijk` and `subtract_independent_offset` was `subtract_offset`, but then I figured writing things out and being quite explicit in function names seemed more idiomatic from what I can tell. Meant I traded one issue for another though.

JoepVanlier commented on docs/examples/fd_fitting_basic/index.rst:
```diff
@@ -0,0 +1,113 @@
+optionally fit multiple models at once. In this example, we only fit a single model::
+
+    M_DNA.load_data(f=f, d=d, name="Twistable WLC")
+    F = pylake.FitObject(M_DNA)
```
Alright. I think I changed it to `snake_case` for variable names everywhere. Classes were already `CapitalCase` afaik.

JoepVanlier commented on lumicks/pylake/fitting/fitdata.py:
```diff
@@ -0,0 +1,103 @@
+        set of transformations from internal model parameters to outer parameters
+    """
+    def __init__(self, name, x, y, transformations):
+        self.x = np.array(x)
+        self.y = np.array(y)
```
Ah thanks, that's good to know. I'll give the whole PR another once over specifically with regard to this next Tuesday just to be sure that I only copy where desired.

JoepVanlier replied:
I've changed most instances of `np.array()` to `np.asarray()`.

dean0x7d commented on docs/api.rst:
```diff
@@ -16,3 +16,17 @@ If you are just looking to get started, read the :doc:`/tutorial/index` first.
     point_scan.PointScan
     correlated_stack.CorrelatedStack
+    fitting.model.Model
+    fitting.Fit
```
This one is a broken link as far as I can see.

dean0x7d commented on docs/examples/fd_fitting_advanced/index.rst:
```diff
@@ -0,0 +1,285 @@
+    figx = 9
+    figy = 6
+
+First we load our data using pylake.
+------------------------------------
```
Periods aren't usually put in titles.

dean0x7d commented on docs/examples/fd_fitting_advanced/index.rst:
```diff
@@ -0,0 +1,285 @@
+    interact(plot_data, data_idx=widgets.IntSlider(min=0, max=len(data) - 1, step=1, value=0))
+
+
+Setting up the global fit.
+--------------------------
```
As above.

JoepVanlier replied:
👍

dean0x7d commented on docs/examples/fd_fitting_advanced/index.rst:
```diff
@@ -0,0 +1,285 @@
+    # Set up the fit, which contains both models
+    fit = pylake.Fit(dna_model, construct_model);
+
+First load the data corresponding to the folded state.
+------------------------------------------------------
```
And here.

dean0x7d commented:
On the topic of API.

I feel like this:
```python
data1 = odijk_with_offsets.load_data(f=f1, d=d1, name="Control")
data2 = odijk_with_offsets.load_data(f=f2, d=d2, name="RecA", DNA_Lc="DNA_Lc_RecA")

odijk_fit = pylake.Fit(odijk_with_offsets)
...plot(odijk_fit[data1], ...)
```

Could be nicer as:
```python
odijk_with_offsets.add_dataset("Control", f1, d1)
odijk_with_offsets.add_dataset("RecA", f2, d2, params={"DNA_Lc": "DNA_Lc_RecA"})

odijk_fit = pylake.Fit(odijk_with_offsets)
...plot(odijk_fit["Control"], ...)
```

Why I think this would be nicer:

* `load_data` sounds like the data is coming *from* somewhere and not like it's going to be added to the model. More specifically, when we talk about `pylake` and loading data, we talk about loading data from a file. So it would be good not to make this ambiguous. As for the concrete naming, I suggested `add_dataset` but I think that any variation on `set` or `add` would be good. Even just plain `add` without the `_dataset` part would be fine.

* `data1` and `name="Control"` are essentially two different ways of assigning a name to the same dataset. If you add the requirement that `name` is unique, then you can use it as a key and there's no need for users to keep track of it in two ways. And then, since it's required and very meaningful, let's put it up at the front.

* Mixing in parameter associations with `**kwargs` can easily result in issues down the road: you may want to add a new argument to the function that conflicts with one of the model parameter names. It's not likely, but as a general rule, I don't think it's nice to mix these two things since one set if very define and the other open-ended.

Other random questions:

* If I read the code correctly, doing `model1 + model2` returns a model without any data. Is this correct? I'm just wondering how manageable that is. And also, users might wonder what's the difference between `pylake.Fit(model1, model2, model3)` vs `pylake.Fit(model1 + model2 + model3)`.

JoepVanlier commented:
> On the topic of API.
> 
> I feel like this:
> 
> ```python
> data1 = odijk_with_offsets.load_data(f=f1, d=d1, name="Control")
> data2 = odijk_with_offsets.load_data(f=f2, d=d2, name="RecA", DNA_Lc="DNA_Lc_RecA")
> 
> odijk_fit = pylake.Fit(odijk_with_offsets)
> ...plot(odijk_fit[data1], ...)
> ```
> 
> Could be nicer as:
> 
> ```python
> odijk_with_offsets.add_dataset("Control", f1, d1)
> odijk_with_offsets.add_dataset("RecA", f2, d2, params={"DNA_Lc": "DNA_Lc_RecA"})
> 
> odijk_fit = pylake.Fit(odijk_with_offsets)
> ...plot(odijk_fit["Control"], ...)
> ```
> 
> Why I think this would be nicer:
> 
> * `load_data` sounds like the data is coming _from_ somewhere and not like it's going to be added to the model. More specifically, when we talk about `pylake` and loading data, we talk about loading data from a file. So it would be good not to make this ambiguous. As for the concrete naming, I suggested `add_dataset` but I think that any variation on `set` or `add` would be good. Even just plain `add` without the `_dataset` part would be fine.
> * `data1` and `name="Control"` are essentially two different ways of assigning a name to the same dataset. If you add the requirement that `name` is unique, then you can use it as a key and there's no need for users to keep track of it in two ways. And then, since it's required and very meaningful, let's put it up at the front.
> * Mixing in parameter associations with `**kwargs` can easily result in issues down the road: you may want to add a new argument to the function that conflicts with one of the model parameter names. It's not likely, but as a general rule, I don't think it's nice to mix these two things since one set if very define and the other open-ended.
> 
> Other random questions:
> 
> * If I read the code correctly, doing `model1 + model2` returns a model without any data. Is this correct? I'm just wondering how manageable that is. And also, users might wonder what's the difference between `pylake.Fit(model1, model2, model3)` vs `pylake.Fit(model1 + model2 + model3)`.

I really like the suggested API changes, but there are a few issues with it. I've renamed `load_data` to `add_data`, no issues there. Made the name first argument and changed out keyworded arguments with `params`, also yes, I agree it is better to just clearly define it as a parameter.

The one that's tricky and makes me doubt is the slicing by the string of the dataset name rather than the handle. It's a fairly big change.

1. This would force us to put `add_dataset` in `Fit`, with a first parameter that refers to the the model we're adding things to, which I have mixed feelings about, because it complicates the more common use case for the more advanced one.
- Dataset name uniqueness cannot be known at the time of constructing the model. It also means that you can't have some very common names in two models. I could see two models having a dataset named `"Control"` for instance.
- One benefit to `add_data` in `Fit` would be that it would also kill the option of having the composite model where both models had data loaded.
- What I don't like about this, is that it makes it non-obvious that `add_data` would be adding data to a model that was passed to `Fit` and that if I'd add that model to another `Fit`, I would have the data added to the earlier `Fit` already added to it.

2. This point is more minor, but as you already pointed out, it would require the data name to be unique. I think this will potentially cause some confusion when people start adding data in loops as they will have to start either string concatenating or using f-strings.

Another alternative would be to just pass the fit to the model, with the data name and have the model query the fit. That is:
```
odijk_model.plot("Control", odijk_fit)
```

I find this a very tricky decision to make to be honest.

dean0x7d commented:
> * One benefit to `add_data` in `Fit` would be that it would also kill the option of having the composite model where both models had data loaded.
> * What I don't like about this, is that it makes it non-obvious that `add_data` would be adding data to a model that was passed to `Fit` and that if I'd add that model to another `Fit`, I would have the data added to the earlier `Fit` already added to it.

I like the idea of adding data to `Fit` instead of `Model`. If I understood your second point, I think that's also solved by making `Model` immutable. So `Fit` would store data associated to a `Model` instead of having `Fit` store `Model`s which store `Data`. Something like this:

```python
model1 = ...
model2 = ...

fit = pl.Fit(model1, model2)
fit[model1].add_dataset("Control", f1, d1)
fit[model1].add_dataset("RecA", f2, d2, params={"DNA_Lc": "DNA_Lc_RecA"})
```

And the simpler single-model case could be:
```python
model = ...
fit = pl.Fit(model)
fit.add_dataset("Control", f1, d1)
fit.add_dataset("RecA", f2, d2, params={"DNA_Lc": "DNA_Lc_RecA"})
```

Here:
```python
class Fit:
    def __init__(self, *models):
        self.models = {id(m): m for m in models}
        self.data = {id(m): Datasets() for m in models}
    
    def __getitem__(self, model):
        return self.data[id(model)]

    def add_dataset(...):
        """Convenience when len(self.models) == 1"""
```

Let me know if you think this is making the API too contrived. The main thing I'd like to avoid is having both the model and fit object be mutable. Especially given that the compatibility of the models is in conflict with the data mutability.

JoepVanlier commented:
It can be done, but it's a big architectural change that worries me a little considering the timeline of things.

> fit[model1].add_dataset("Control", f1, d1)
> fit[model1].add_dataset("RecA", f2, d2, params={"DNA_Lc": "DNA_Lc_RecA"})

While I do like what this API looks like, I see complications in `fit[model1]` returning a `Dataset`. To be able to add a new set of data, I need to know the model. This is because I think that validating that the transformation the user is making is valid is critical. Invalid parameter transformations can lead to subtle mistakes that may not be detected otherwise. Of course, I could still throw at a later stage when the user tries to fit, but I imagine deferring the exception of an invalid parameter transform in a dataset that was added until the model is being built (essentially when the fitting is going to happen) is going to be kind of a nuisance. You'd be adding ten datasets and the only at the fitting stage suddenly discover that the first one had a mistake somewhere, and now you have to start over.

That leaves me with two options, either I put a reference to `Model` in `Dataset`, which is possible, but it smells and raises the question of whether data shouldn't be in the model in the first place. Or the call becomes:

`fit.add_data(model_ref, "Dataset name", ...)`

And have fit handle passing the model. It does mean more call arguments though, and no nice convenience function for small models.

JoepVanlier commented:
Changes:
- New API implemented.
- Examples have been moved to this new API.
- Tutorial has been moved to this new API.
- Some coverage gaps filled.

dean0x7d commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,343 @@
+    
+We see that indeed the second parameter now appears. We also note that the parameters
+from the first fit changed. If this was not intentional, we should have fixed
+these parameters after the first fit. For example, we can fix the parameter `DNA/Lp`
+by invoking::
+
+    >>> odijk_fit["DNA/Lp"].vary = False
```
Not a big deal, but consider inverting this logic to `.fixed = True`. I just find it nicer to instruct someone to "Do `x` by setting `x = True`" rather than "Do `x` by setting `y = False`".

JoepVanlier replied:
I actually had a preference for `.fixed` as well. The reason I went with `vary` was to be compatible with `lmfit` in that regard as quite a few people might already be familiar with that library. Then again, we do quite different things in many places from `lmfit`, so we might as well switch to `.fixed`. I'll change it.

JoepVanlier commented on docs/api.rst:
```diff
@@ -16,3 +16,17 @@ If you are just looking to get started, read the :doc:`/tutorial/index` first.
     scan.Scan
     point_scan.PointScan
     correlated_stack.CorrelatedStack
+    fitting.model.Model
+    FdFit
+
+    force_offset
+    distance_offset
+    marko_siggia_ewlc_force
+    marko_siggia_ewlc_distance
+    marko_siggia_simplified
+    odijk
+    inverted_odijk
+    freely_jointed_chain
+    inverted_freely_jointed_chain
+    twistable_wlc
+    inverted_twistable_wlc
```
👍 Fully agree.

Much more structured that way.

JoepVanlier commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,343 @@
+.. warning::
+    This is beta functionality. While usable, this is a beta-functionality which
+    has not yet been tested in a sufficient number of different scenarios. The API
+    may also still be subject to change.
```
👍

JoepVanlier commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,343 @@
+extensible worm-like-chain model by Odijk et al. Let's have a look at it. We can construct
+this model using the supplied function `pylake.odijk()`.
+
+Note that we also have to give it a name. This name will be prefixed to model specific
+parameters in this model. So for instance, the persistence length in this model, would be
+named `DNA/Lp`. Let's have a look at the model and its parameters::
+
+    >>> odijk_forward = pylake.odijk("DNA")
+    >>> odijk_forward
```
- That's a really good point.
- I will move the DNA/Lp to after the code blurb. That way the model is introduced first.
```
Note how most of the parameters listed have the model name prefixed. For
example, the persistence length in this model is now named `DNA/Lp`. Here
`DNA` refers to the name we gave to the model.

One parameter stands out, which is kT as it doesn't have `DNA` prefixed. The
reason for this is that this parameter is not model specific (in this case the
parameter represents the Boltzmann constant times the temperature).
```

- I'd rather keep it `>>> model` as that gives the html representation (with tex equation) in Jupyter, while print gives a plaintext one.

```
Entering model prints the model equation and its default parameters::
```

JoepVanlier commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,343 @@
+we have the inverted Odijk model::
+
+
+    odijk_inverted = pylake.inverted_odijk("DNA")
```
👍

JoepVanlier commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,343 @@
+
+We can have a quick look at what this model looks like and which parameters are in there::
+
+    >>> pylake.inverted_odijk("DNA")
```
Agree with the reuse, prefer to not use print because of the html repl thing. So would suggest `>>> model`

JoepVanlier commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,343 @@
+This seems to be what we want. Force as a function of distance. Now let's say we have acquired data,
+but we notice that the template matching wasn't completely optimal. And that we had some force drift,
+while forgetting to reset the force back to zero. In this case, we can incorporate offsets in our
+model. We can introduce an offset in the independent parameter, by calling `subtract_independent_offset`
```
👍

dean0x7d commented on docs/examples/fd_fitting_basic/index.rst:
```diff
@@ -0,0 +1,194 @@
+provides a lot of additional freedom in the model::
+
+    fit_odijk.add_data("Inverted Odijk", force, distance)
+    fit_odijk["DNA/d_offset"].upper_bound = .01
+    fit_odijk["DNA/d_offset"].lower_bound = -.01
```
Maybe add the leading `0` since `-.01` looks a bit strange and could distract from the main point: https://herbsutter.com/2020/05/19/of-feedback-and-little-things/

JoepVanlier replied:
Nice read :)

Changed!

dean0x7d commented on docs/examples/fd_fitting_basic/index.rst:
```diff
@@ -0,0 +1,194 @@
+Again, we incorporate an offset in both distance and force to compensate for small
+offsets that may exist in the data:::
+
+    m_dna = pylake.inverted_twistable_wlc('DNA').subtract_independent_offset() + pylake.force_offset('DNA')
```
Again, make sure it fits without horizontal scrolling. Line breaks are fine.

JoepVanlier replied:
👍

dean0x7d commented on docs/examples/fd_fitting_basic/index.rst:
```diff
@@ -0,0 +1,194 @@
+
+We could add more datasets in a similar manner, but in this example, we only fit a single model. Let’s load the
+parameters from our previous fit to use them as initial guesses for this one. We also fix the twist rigidity and
+critical force values to values from literature (analogous to Broekmans et al. "DNA twist stability changes with
```
```suggestion
critical force to values from literature (analogous to Broekmans et al. "DNA twist stability changes with
```

JoepVanlier replied:
Oops! Thanks.

JoepVanlier commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,343 @@
+If we also expect an offset in the dependent parameter, we can simply add an offset model to our
+model::
+
+    odijk_with_offsets = pylake.inverted_odijk("DNA").subtract_independent_offset() + pylake.force_offset("DNA")
```
Regarding point 1 and 2:

This was a tough call for me too. I'm not a huge fan of the `.subtract_independent_offset()` either. I don't think there's a reason the addition operator couldn't be implemented for a distance offset, but it could generate some potentially unsafe things.

The way it's implemented now, the plus operator behaves like you'd expect from a plus operator i.e. `model_f() + model_g() + model_h()` => `f(x) + g(x) + h(x)`. Addition only being allowed for matching independent variables.

I'd find it very strange and unexpected if for one particular model it'd suddenly behave as
`model_f() + model_g() + model_h()` => `inv(inv(f)(y) + g(y)) + h(x)`.

Note that explicitly doing the inversions like `inv(inv(f)(y) + g(y)) + h(x)` is also quite slow, subtracting an offset is a special case that I treat with a specialized model that propagates sensitivities for that case specifically. Since it is by far the most common use case I feel that that specialization is justified since it is really about 20x-50x faster.

If you do want to implement it by operators you would have to be explicit now, i.e.:
```python
(lk.odijk("DNA") + lk.distance_offset("DNA")).invert() + force_offset("DNA")
```
(Note that I used the forward odijk now).

I figured it'd be much safer to be specific about only allowing addition of models that have matching independent and dependent variables, and not do implicit inversions. I figured keeping the API constrained in this way for now would prevent surprises that I may not yet be able to foresee.

In addition to that an offset is special in the sense that you can easily propagate it cheaply and it is the most common use case for this purpose.

Regarding point 3, at the moment no. Whereas at the moment, I'd say it makes sense to enforce the independent offset to be named the same, as it is really tied to that submodel whereas the offset is kind of outside of the other models in the construct. Let me illustrate that with an example:

```python
protein(d).subtract_independent_offset() + dna(d).subtract_independent_offset() + lk.force_offset("construct")`
Note how the offsets for the protein and dna are really subtracting the distance that goes into that model specifically, whereas the force offset applies to both.
```

dean0x7d replied:
That all makes sense. Thanks! It would be useful to add some of that explanation to the tutorial. It might be too much to dump on readers upfront so consider restructuring the tutorial a little bit. Right now, it's divided into sections by stages of the fitting process and the entire tutorial runs through one fit, introducing more advanced concepts inline in every section. It might be nice to restructure it so that it runs through a very basic fit from start to finish and then does on to introduce more advanced concepts like composite models, inversion, and the related constraints.

dean0x7d commented on docs/examples/fd_fitting_basic/index.rst:
```diff
@@ -0,0 +1,194 @@
+critical force values to values from literature (analogous to Broekmans et al. "DNA twist stability changes with
+magnesium (2+) concentration." Physical review letters 116.25 (2016))::
+
+    fit_twlc << fit_odijk
```
I guess that this means "load parameters from our previous fit" but it's far removed from the text. I would strongly recommend making this a named method instead of the `<<` operator. I'm very much for operator overloading in cases of frequently used constructs. It can make things easier to read, especially for chaining. But here I don't see an obvious advantage, especially since I see it being used only once in all the examples. It takes more time to explain than to use if it's used very rarely.

JoepVanlier replied:
Yeah, that's fair. How about `load_parameters_from`?

`load_parameters` makes me think it'll overwrite what parameters are in the list (the keys as well as the values) whereas this only assigns to keys that already exist in the target and doesn't require that all keys are even present.

dean0x7d replied:
Maybe something closer to the intention? `.set_initial_guess()`? In any case, avoid the word `load` as mentioned before in the context of setting datasets.

JoepVanlier replied:
Hmm, I'm not sure. It has more uses than just initial guesses though. In some cases, you may want to load them but fix them, when you fix a DNA linker for instance. In that case it would hardly function as an initial guess. Initial guess also makes me think that it would have to give a complete parameter vector, while this is typically used to provide values for a subset.

I'd prefer something closer to what it does like `set_parameters`. The reason I preferred `load` over `set` was because `set` sounded a bit like it would override the complete vector (adding ones that weren't in the target for instance). `import_parameter_values` maybe? I don't know. I'll sleep on it! :D

dean0x7d replied:
`.update_params()`? Update seems to be the thing that's happening.

* I also noticed that `parameters` vs `params` in the API. Would be good to unify everything. My vote is for `params` since it's shorter :)

JoepVanlier replied:
👍

JoepVanlier replied:
`params` it is. Also, `update_params()` sounds good. 👍

JoepVanlier replied:
`params` and `.update_params()` it is! :)

dean0x7d commented:
Docs file structure comments:
* Try to name the `.rst` files so that they are related to the page title. Right now you have `Twistable Worm-Like-Chain Fitting` in a file called `fd_fitting_basic`. Something like `twlc_fitting` might be a better name. These names end up in the URL so it can be confusing if there is a mismatch.
* `fd_fitting_advanced` seems to be in the file structure but it's not in the `toctree` so it's not rendered in the HTML. Is this intentional?

JoepVanlier commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,343 @@
+    two_odijk = (pylake.odijk("DNA") + pylake.odijk("protein") + pylake.distance_offset("offset")).invert()
+
+
+Note how we added three models and then inverted the composition of those models. The parentheses 
```
👍

JoepVanlier commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,343 @@
+Adding data to the fit
+----------------------
+
+Next up, is adding some data. Let's assume we have two data sets. One was acquired in the presence
```
👍

JoepVanlier commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,343 @@
+
+    odijk_fit = pylake.FdFit(odijk_with_offsets)
+
+Adding data to the fit
+----------------------
```
👍 Good point! RST Syntax checks out btw.

JoepVanlier commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,343 @@
+of a ligand, and another was measured without a ligand. We expect this ligand to only affect the
+contour length of our DNA. Adding the first data set to the fit is simple::
+
+    odijk_fit.add_data("Control", force1, distance1)
+
+For the second data set, we want the contour length to be different. We can achieve this by renaming
```
Good point 👍

dean0x7d commented on docs/examples/fd_fitting_intermediate/index.rst:
```diff
@@ -0,0 +1,374 @@
+
+Now comes the more challenging part. Inverting the model for contour length. Luckily, this
+procedure has already been implemented in Pylake. The function `parameter_trace` inverts the
+model for a particular model parameter. Let's have a look at the parameters it needs::
+
+    >>> help(pylake.parameter_trace)
+
+    Help on function parameter_trace in module lumicks.pylake.fitting.detail.parameter_trace:
+
+    parameter_trace(model, parameters, inverted_parameter, independent, dependent, **kwargs)
```
Rather than printing the docstring here, I would link to the API reference for that function. This print is essentially just a copy/paste of the docstring and that means that you need to maintain it in two different places.

JoepVanlier replied:
Switched it to a reference :)

dean0x7d commented on docs/examples/fd_fitting_intermediate/index.rst:
```diff
@@ -0,0 +1,374 @@
+The parameter we wish to invert for is `DNA/Lc` and for the independent and dependent variables we simply
+pass the dataset::
+
+    Lcs = pylake.parameter_trace(m_dna, fit["RecA"], 'DNA/Lc', distance_dynamic, force_dynamic)
```
It would be good to keep the string style consistent `""` vs `''`. My preference is usually `""` since it makes it easier to manage moving between Python and C++, but either is fine as long as it's used consistently throughout the docs. Don't worry about updating everything. Right now, it would just be good to have a single style on a single line of code :)

JoepVanlier replied:
Whoops. Not sure how that even happened. 😊

dean0x7d commented on docs/examples/fd_fitting_intermediate/index.rst:
```diff
@@ -0,0 +1,374 @@
+.. image:: output_10_8.png
+
+Here we can see the different contour length transitions quite clearly. There seems to be one region
+of contour lengths around 3.2 before finally lengthening to 3.4 micrometers.
```
Comment for the overall page: I'm curious about what BD will say, but to me this looks like a very nice example. Well written and it nicely walks the user through the fitting process and various data-specific challenges.

JoepVanlier replied:
I have just added the suggestions from BD to the documentation. Tomorrow I will try to clean up the history a bit and undo the FD => Fd bit for `FDCurve` and `DownsampledFD`.

I'd prefer keeping the `import lumicks.pylake as lk` for the examples, since they share the same header. How do you feel about this?

So how strongly do you think this commit history should be squashed?

On March 24/25 I reconstructed the history up to that point and made the history fairly clean (even though it is a big chunk of commits). Shall I clean up the history post 24th grouping the changes?

dean0x7d commented on lumicks/pylake/__init__.py:
```diff
@@ -2,7 +2,10 @@
+from .fitting.models import *
 from .correlated_stack import CorrelatedStack
+from .fitting.fit import FdFit
+from .fitting.detail.parameter_trace import parameter_trace
```
This probably doesn't belong in `detail` if it's being made available as public API anyway.

JoepVanlier replied:
Ah yes. Originally, I was planning to make it accessible through `FdFit`, but I think a free function is better now. I'll move it!

dean0x7d commented on lumicks/pylake/detail/utilities.py:
```diff
@@ -1,3 +1,7 @@
+import numpy as np
+import matplotlib.pyplot as plt
+from matplotlib import colors
```
Prefer `import matplotlib as mpl` and then `mpl.colors`.

JoepVanlier replied:
👍

dean0x7d commented on lumicks/pylake/fitting/models.py:
```diff
@@ -0,0 +1,390 @@
+
+
+def marko_siggia_ewlc_force(name):
+    """
+    Marko Siggia's Worm-like Chain model with force as dependent parameter.
```
```suggestion
    """Marko Siggia's Worm-like Chain model with force as dependent parameter.
```

JoepVanlier replied:
👍

JoepVanlier commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,343 @@
+    odijk_fit["DNA/Lp"].lower_bound = 39
+    odijk_fit["DNA/Lp"].upper_bound = 80
+
+After this, the model is ready to be fitted::
+
+    odijk_fit.fit()
+
+Note that multiple models can be fit at once, by just supplying more than one model::
```
Ah yes, I should elaborate more. As far as I know, all functions of the `FdFit` object should still just work. You can just `plot` a `FdFit` that wasn't fitted for instance. It just fits the data poorly.  I do want to keep the ability to plot it before fitting as this can be useful for finding a good initial guess.

Since parameter estimation by convex optimization is often quite dependent on the initial guess, one might want to play with parameters first before actually fitting the model.

I changed the description to:
```rst
After this, the model is ready to be fitted. We can fit the model to the data by calling
the function `.fit()`. This estimates the model parameters by minimizing the least squares
differences between the model's dependent variable and the data in the fit::

    fit.fit()

After this call, the parameters will have new values that should bring the model closer
to the data. 
```

dean0x7d replied:
Cool, sounds good. Maybe add a bit of the explanation from the comment as well. E.g. it's nice to know that you can plot before fitting and that there's a purpose for that.

JoepVanlier replied:
Alright. Added the following to the plotting section:

```rst
A model can be plotted before it is fitted. This can be useful when the default parameter
values don't seem to work very well. Parameter estimation is typically initiated from
an initial guess. A poor initial guess can lead to a poor parameter estimate. Therefore, 
you might want to see what your initial model curve looks like and set some better
initial guesses yourself when you run into trouble.
```

As it felt a bit more natural there. I'll have one more once-over before committing.

JoepVanlier commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,343 @@
+Global fits versus single fits
+------------------------------
+
+The `Fit` object manages a fit. To illustrate its use, and how a global fit differs from a
```
Correct. That one must have slipped through the cracks, thanks.

`Fit` still exists. `FdFit` is a specialized one that has a custom `add_data` that makes sure the order is always `f`, `d` irrespective of whether you have a forward or inverse model (since that lead to confusions otherwise).

In addition, the specialization is nice to have for when we actually couple this more directly to the `FdCurve` class, as we can then keep the `Fd`-specific stuff contained in its own class. I could see reusing some of this more general infrastructure for other parameter estimation problems for instance.

JoepVanlier commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,343 @@
+        odijk_fit.fit()
+        print(odijk_fit["DNA/Lc"])
+
+The difference between these two is that the former sets up a single model, that has to fit
+all the data whereas the latter fits all the data sets independently. The former has one single
```
👍

JoepVanlier commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,343 @@
+complicated. For instance, we may have two worm like chain models that we wish to add, and then invert.
+For the Odijk model, this can be done as follows::
+
+    two_odijk = (pylake.odijk("DNA") + pylake.odijk("protein") + pylake.distance_offset("offset")).invert()
```
👍

JoepVanlier commented on docs/examples/fd_fitting_basic/index.rst:
```diff
@@ -0,0 +1,194 @@
+    import matplotlib.pyplot as plt
+    from lumicks import pylake
+
+Load the data from disk
+-----------------------
```
👍

JoepVanlier commented on docs/examples/fd_fitting_basic/index.rst:
```diff
@@ -0,0 +1,194 @@
+--------------------------
+
+We clearly see that the force starts levelling out at high forces in the data. This
+is a clear sign that to be able to describe this data, we'll need something more complex
+in order to capture this behaviour we see in the data. The twistable worm like chain
```
True, thanks! I wonder if I somehow mangled that sentence by editing it and then just kept reading over it.

JoepVanlier commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,343 @@
+the inversion is done numerically, rather than analytically. For more complex examples on how this
+inversion may be used, please see the examples.
+
+For a full list of models that are available, please refer to the documentation by invoking
+`help(pylake.fitting.models)`.
```
👍

JoepVanlier commented on docs/examples/fd_fitting_intermediate/index.rst:
```diff
@@ -0,0 +1,374 @@
+to the model name slash the new parameter name. Let's rename the contour length Lc, persistence
+length Lp and stiffness St for this data set::
+
+    data2 = fit.add_data("RecA", force_reca, distance_reca, params={"DNA/Lc": "DNA/Lc_RecA", "DNA/Lp": "DNA/Lp_RecA",
```
👍

JoepVanlier commented on docs/examples/fd_fitting_basic/index.rst:
```diff
@@ -0,0 +1,194 @@
+    fit_odijk = pylake.FdFit(m_odijk)
+
+Considering that this model only describes the lower part of the curve, we have to
+extract the data that is relevant to us. We can obtain this data from the ``FdCurve``
```
I asked BD, and Fd seems to be more common. I'll switch FdCurve to be consistent.

dean0x7d replied:
That's fine, just name sure to note `FDCurve` -> `FdCurve` as breaking change in the changelog and the version must go to v0.5.0.

JoepVanlier commented on docs/tutorial/fdfitting.rst:
```diff
@@ -0,0 +1,343 @@
+calling plot on the model directly::
+
+    odijk_with_offsets.plot(odijk_fit['Control'], np.arange(2.0, 5.0, .01), fmt='k--')
+    odijk_with_offsets.plot(odijk_fit['RecA'], np.arange(2.0, 5.0, .01), fmt='k--')
```
You raise a good point. There should probably be an easier way for users to plot things. So... the new one:

If you have just one model: `fit.plot()` will take care of everything.

If you have more than one: `fit[model1].plot()` will plot all data corresponding to model 1. I prefer the angle brackets because it is analogous to the API elsewhere.

And you can (if you want) also plot a specific dataset: `fit.plot("data name")`. Or in the multimodel case `fit[model1].plot("data name")`

You can also pass an optional `independent` if you want to plot over a custom time range.

I think this covers all the cases that I've used in the notebooks practically, with relative ease.

I do think the model.plot will have a role for sensitivity analyses and the like, and it is called by `FdFit.plot` now too, so I would like to keep that one around too, just not make it the goto one.

dean0x7d replied:
That's a great solution! I really like it!

JoepVanlier replied:
Cool, thanks! 👍

JoepVanlier commented on docs/examples/fd_fitting_intermediate/index.rst:
```diff
@@ -0,0 +1,374 @@
+Let’s print the contour length difference due to RecA. We multiply by 1000 since we desire this value in
+nanometers::
+
+    >>> print(f"Contour length difference: {(fit['DNA/Lc_RecA'].value - fit['DNA/Lc'].value) * 1000:.2f} [nm]")
```
Yes, I will go over all the code parts once more at the very end to make sure it fits.

The only place where I can't guarantee no hori scrollbar is printed output. It might be a nice future improvement to have "nice" line-breaking of equations, but this is a pretty hard problem.

dean0x7d replied:
Yeah, don't worry about the printed output.

JoepVanlier replied:
👍

JoepVanlier commented on docs/examples/fd_fitting_intermediate/index.rst:
```diff
@@ -0,0 +1,374 @@
+    marko_siggia_fit.add_data("Control", force_control, distance_control)
+    marko_siggia_fit.add_data("RecA", force_reca, distance_reca, params={"DNA/Lc": "DNA/Lc_RecA", "DNA/Lp": "DNA/Lp_RecA",
+                                                                         "DNA/St": "DNA/St_RecA"})
+    marko_siggia_fit.fit();
```
👍

JoepVanlier commented on docs/examples/fd_fitting_intermediate/index.rst:
```diff
@@ -0,0 +1,374 @@
+changes, let's have a look at what these models predict for the change in contour length::
+
+    >>> print(f"Contour length difference Odijk: {(fit['DNA/Lc_RecA'].value - fit['DNA/Lc'].value) * 1000:.2f} [nm]")
+    >>> print(f"Contour length difference Marko-Siggia: {(marko_siggia_fit['DNA/Lc_RecA'].value - marko_siggia_fit['DNA/Lc'].value) * 1000:.2f} [nm]")
```
👍

JoepVanlier commented on docs/examples/fd_fitting_basic/index.rst:
```diff
@@ -0,0 +1,194 @@
+We could add more datasets in a similar manner, but in this example, we only fit a single model. Let’s load the
+parameters from our previous fit to use them as initial guesses for this one. We also fix the twist rigidity and
+critical force values to values from literature (analogous to Broekmans et al. "DNA twist stability changes with
+magnesium (2+) concentration." Physical review letters 116.25 (2016))::
```
😁

> Docs file structure comments:
> 
> * Try to name the `.rst` files so that they are related to the page title. Right now you have `Twistable Worm-Like-Chain Fitting` in a file called `fd_fitting_basic`. Something like `twlc_fitting` might be a better name. These names end up in the URL so it can be confusing if there is a mismatch.

Renamed. Actually, this was a bit of a pet peeve to begin with, because one annoying thing was that if you downloaded the file as jupyter notebook, it would be named index.rst which is a annoying. I renamed the other examples too.

> * `fd_fitting_advanced` seems to be in the file structure but it's not in the `toctree` so it's not rendered in the HTML. Is this intentional?

I removed it for now. While that notebook would be great to have in the future I think, it depends on some yet unreleased piezo tracking code. I have a reasonably clean implementation of that, but we should discuss on how to review that, and whether we also want that in Pylake. Until that time, I will just leave the notebook out.

JoepVanlier commented on docs/examples/fd_fitting_basic/index.rst:
```diff
@@ -0,0 +1,194 @@
+
+Let's set up the Odijk model and create the fit::
+
+    m_odijk = pylake.inverted_odijk('DNA').subtract_independent_offset() + pylake.force_offset('DNA')
```
👍

dean0x7d commented:
This PR is superseded by https://github.com/lumicks/pylake/pull/33 with a cleaned-up git history.
